### DISCUSSION TITLE
custom installer... `getInstallPath()`

### DISCUSSION THREAD

**Discussion by staabm** [composer/composer#10523] [Q&A / Support] [answered]
I have implemented a custom-installer lib which worked for us for a few years:
https://github.com/staabm/complex-library-installer/blob/master/src/LibraryInstaller.php#L13

recently we see errors in combination with the phpstan-extension installer:
https://github.com/phpstan/extension-installer/issues/55

I guess the underlying problem is, that my custom installer returns relativ paths in
https://github.com/staabm/complex-library-installer/blob/396ab67a9d4dfd67542fc8fbe4b91e60b9922013/src/LibraryInstaller.php#L49

but the phpstan-extension installer assumes paths returned by `getInstallPath()` are absolute:
https://github.com/phpstan/extension-installer/blob/994598bb7a3df21d48510dbd358973350e66b9ea/src/Plugin.php#L117

the `InstallerInterface` does not provide a hint, whether the path returned needs to be a absolute one:
https://github.com/composer/composer/blob/a0b0ddcd73f3f5a19d7028063877efd97249f22a/src/Composer/Installer/InstallerInterface.php#L113-L119

could someone clarify whether my installer should always return a absolute path, or alternatively whether the phpstan-extension installer should try to not make the assumption that `$installationManager->getInstallPath($package);` will be a absolute path?

----

within composer I found 1 spot where the install-path is passed to `realpath`, which might be an indicator that the path is allowed to be relative:
https://github.com/composer/composer/blob/a0b0ddcd73f3f5a19d7028063877efd97249f22a/src/Composer/Package/Locker.php#L486

**Comment by Seldaek** [comment, marked as answer]
It is allowed to be relative for historical reasons and was notably done like this in composer/installers until 2.0 (https://github.com/composer/installers/commit/092ce6f9b753d7a4ca6a9fe373ac63c42b4a6aa8). Ideally though IMO returning absolute paths is better and safer, because within Composer itself it's always absolute and thus many things make that assumption and when returning relative paths you risk finding edge cases ;)

Feel free to send a PR documenting this better in getInstallPath :)

**Reply by staabm** [reply]
thx for the info, doc change incoming in https://github.com/composer/composer/pull/10524
